tt_eager/tt_dnn/op_library/prod/kernels/reader_prod_nc.cpp

```diff
@@ -0,0 +1,59 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cstdint>
+
+#include "dataflow_api.h"
+#include "tt_eager/tt_dnn/kernels/dataflow/moreh_common.hpp"
+
+inline uint32_t get_read_tile_id(uint32_t tile_id, uint32_t dim, uint32_t input_tile_offset, uint32_t HtWt) {
+    return (dim == 0 ) ? (tile_id) : (tile_id / HtWt * input_tile_offset) + (tile_id % HtWt);
+}
+
```

Comment:
Division and modulo are expensive operations on the riscs. You should be able to get the id without this by just incrementing/using ifs in the loop. If dim is passed as a compile time arg you can also constexpr it so it isn't evaluated at runtime

Reply:
Will replace  `return (dim == 0 ) ? (tile_id) : (tile_id / HtWt * input_tile_offset) + (tile_id % HtWt);` with
```
    if(dim == 0){
        return tile_id;
    }else {
        uint32_t a = 0 ;
        while (tile_id >= HtWt){
            tile_id-= HtWt;
            a = a + 1;
        }
        uint32_t b = 0;
        for (uint32_t i = 0; i < input_tile_offset; ++i) {
            b = b + a;
        }
        return b + tile_id;
    }
```